diags/zpool-properties.t

```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+# check that ZFS pool properties are set as specified
+
+export PATH=/bin:/usr/bin:/sbin:/usr/sbin
+
+declare -r description="Check zfs pool properties"
+
+# Source functions
+source ${NODEDIAGDIR:-/etc/nodediag.d}/functions-tap || exit 1
+
+diag_handle_args "$@"
+
+pools=`zpool list -H | awk '{print $1}'`
+[ -z "$pools" ] &&  diag_plan_skip "no ZFS pools" >&2
+
+function verify_property {
+	local fname="verify_property"
+	local pool=$1
+	local propname=$2
+	local expectedval=$3
+
+	if [ -z "$pool" ]; then
+		echo "BUG: $fname missing argument pool" >&2
+		exit 1
+	fi
+
+	if [ -z "$propname" ]; then
+		echo "BUG: $fname missing argument propname" >&2
+		exit 1
+	fi
+
+	if [ -z "$expectedval" ]; then
+		echo "BUG: $fname missing argument expectedval" >&2
+		exit 1
+	fi
+
+	propval=`zpool list -H -o ${propname} ${pool} 2>/dev/null`
+	if [ "${propval}" != ${expectedval} ] ; then
+		diag_fail "pool ${pool} property ${propname} is ${propval}, expected ${expectedval}" >&2
+	else
+		diag_ok "pool ${pool} property ${propname} is ${propval}" >&2
+	fi
+}
+
+function handle_test {
+	local fname="handle_test"
+	local pool=$1
+	local propname=$2
+	local expectedval=$3
+	local action=$4
+	
+	if [ -z "$action" ]; then
+		echo "BUG: $fname missing argument action" >&2
+		exit 1
+	fi
+
+	if [ "$action" = "count_tests" ] ; then
+		num_tests=$((num_tests+1))
+	else
+		verify_property ${pool} ${propname} ${expectedval}
+	fi
+}
+
+function foreach_pool_and_property {
+	local action=$1
+
+	for pool in ${pools}
+	do
+		for idx in ${!DIAG_ZPOOL_NAME[@]}
+		do
+			poolregex=${DIAG_ZPOOL_NAME[$idx]}
+
+			if [[ ! ${pool} =~ ${poolregex} ]]; then
+				continue
+			fi
+
+			handle_test ${pool} autoreplace ${DIAG_ZPOOL_AUTOREPLACE[$idx]} $action
+			handle_test ${pool} multihost   ${DIAG_ZPOOL_MULTIHOST[$idx]}   $action
+		done
+	done
+}
+
+#
+# number of tests
+#
+num_tests=0
+foreach_pool_and_property count_tests
```

Comment:
This appears to make the action (arg $4) always be count_tests, which results in no actual comparison happening between the set value and the expected value.

Reply:
Correct.  With action="count_tests", the tests are not executed.  We just calculate the number of tests.

If you look at line 89, you see that we have to declare, in advance, the number of tests which will be executed.  But the syntax supports different sets of tests, and so the number can only be determined by counting.  For example:
```
DIAG_ZPOOL_NAME[0]=".*/mdt1"
DIAG_ZPOOL_MULTIHOST[0]="on"

DIAG_ZPOOL_NAME[0]=".*/ost1"
DIAG_ZPOOL_MULTIHOST[0]="on"
DIAG_ZPOOL_AUTOREPLACE[0]="on"
```

If run on a jet MDS, with only one pool imported, there would only be 1 test.  But if run on a jet OSS with both pools imported, there would be 4 tests.

So with zpool-properties.t, I handle that with a common function that loops over pools foreach_pool_and_property.  And if action is "count_tests" then the expected/actual values aren't compared, num_tests is incremented.  If action is anything else, then expected/actual values are compared.

Reply:
Does that change your answer, about which of these two versions you prefer?  This is the key difference.

Reply:
No, I don't think it changes my preference, though perhaps I'm not clear of the scope of this patch. In production, when wanted to verify pool-level attributes, what would you be setting the action to? Just something else?

Reply:
If you look at line 95, foreach_pool_and_property is called with argument "do_tests".  This results in the actual values being compared with the expected ones.

This file (zpool-properties.t) calls foreach_pool_and_property twice.  First it counts the tests that will be done (line 87) and then declares that number (line 89) and finally actually performs the tests and complains if appropriate (line 95).

Reply:
Oh dear. I had copied and pasted the script to test on zinc and somehow that last line got missed, thus leading to a lot of head scratching. It now makes much more sense.

Reply:
No problem.  I'm going to re-write zfs-properties to use the same structure, and then put both of them on jet (where it should test stuff) and opal (where it shouldn't break anything).

Reply:
And in fact I see a bug in zpool-properties.t that I need to fix...

Reply:
zfs-properties.t is rewritten, and the bugs I found are fixed.